### ISSUE TITLE
Sidekiq in development doesn't respect initializer config

### ISSUE THREAD

**Issue by sj26** [DataDog/dd-trace-rb#80]
We have configured datadog tracer to be enabled only in production:

```ruby
# config/initializers/datadog-trace.rb
Rails.configuration.datadog_trace = {
  enabled: Rails.env.production?,
}
Datadog::Tracer.log.debug("Rails.configuration.datadog_trace = #{Rails.configuration.datadog_trace.inspect}")
```

But when booting sidekiq in development, datadog tracer starts recording and attempting to write traces much earlier when sidekiq communicates with redis during startup (after the railtie's before_configuration but before after_initialize):

```
$ bin/sidekiq
xD, [2017-03-03T00:43:26.295524 #75633] DEBUG -- : using Msgpack encoder
D, [2017-03-03T00:43:27.920030 #75633] DEBUG -- : 'redis-rails' module not found, Datadog 'redis-rails' integration is disabled
D, [2017-03-03T00:44:28.08712 #75633] DEBUG -- : Rails.configuration.datadog_trace = {:enabled=>false}
2017-03-03T00:43:28.142Z 75633 TID-oxhxghe0g INFO: Booting Sidekiq 4.2.9 with redis options {:url=>"redis://localhost:6379/1", :namespace=>"sidekiq"}
D, [2017-03-03T00:43:28.148760 #75633] DEBUG -- : Set pin redis on Redis::Client.
D, [2017-03-03T00:43:28.149140 #75633] DEBUG -- : Get pin from Redis::Client.
D, [2017-03-03T00:43:28.152427 #75633] DEBUG -- : Get pin from Redis::Client.
D, [2017-03-03T00:43:28.152869 #75633] DEBUG -- : Writing 2 spans (enabled: true)
# ... a bunch of other boot stuff
# I added this log statement to the beginning of Datadog::Railtie#after_initialize method for tracing:
D, [2017-03-03T00:43:28.349357 #75633] DEBUG -- : Datadog::Railtie#after_initialize
# (the sidekiq banner and "starting processing" below here)
# now a bunch of:
E, [2017-03-03T00:43:57.192496 #75633] ERROR -- : Failed to open TCP connection to localhost:7777 (Connection refused - connect(2) for "localhost" port 7777)
```

Perhaps when setting `Datadog::Tracer.enabled = false` the tracer should also discard any buffered traces completely? Ideally when booting within rails the tracer should only buffer traces and not attempt to send them until after rails has finished booting.

**Comment by sj26** [comment]
The only way to prevent this seems to be preventing `ddtrace` from being required at all outside production by using `gem "ddtrace", group: "production"` in the Gemfile. We'd prefer to keep it loaded in all environments for parity with production and so we can add application instrumentation.

**Comment by sj26** [comment]
More specifically, this seems to be the problem:

https://github.com/DataDog/dd-trace-rb/blob/10e4534db082d8787382dd54126560f72dcd4a76/lib/ddtrace.rb#L48

Because redis is patched and starts writing spans well before anything in `config/initializers/*` is loaded there are already recorded spans attempting to be sent to the datadog agent before the tracer is disabled in development.

**Comment by palazzem** [comment]
Thank you for reporting that @sj26 ! we'll be sure that the next release doesn't have such kind of issues.

**Comment by palazzem** [comment]
https://github.com/DataDog/dd-trace-rb/pull/86 should remove the root cause about Redis sending traces despite defined settings. A configuration like:
```ruby
Rails.configuration.datadog_trace =
  {
    # may depend by your conditional env
    enabled: false,
    # these may be 'true' to keep parity with your production
    auto_instrument: true,
    auto_instrument_redis: true
  }
```
ensures that:
* Rails is auto-instrumented after bundler requires `ddtrace`
* Redis is auto-instrumented after bundler requires `ddtrace`
* if `redis-rails` is used and Redis is auto-instrumented, `redis-rails` will use the traced Redis client
* disabling the tracer means that traces will not be reported but the code will still be instrumented like in your production environment.

**Comment by palazzem** [comment]
Fixed with #86 . Redis auto-instrumentation (available after any kind of `require 'ddtrace'`) honors the Rails initializer. So if the tracer is disabled, Redis is instrumented after so that no traces are sent from the transport layer (but traces are still generated and Redis is still instrumented). This plays well with Rake scripts and Sidekiq (in environments where the tracer is not enabled).

If you spot another problem, feel free to re-open the issue or to create a new one! The next release will be available by the end of the week and it includes a better Sidekiq support!
